nav2_collision_monitor/src/polygon_base.cpp

```diff
@@ -0,0 +1,175 @@
+      node->get_parameter(polygon_name_ + ".action_type").as_string();
+    if (at_str == "stop") {
+      action_type_ = STOP;
+    } else if (at_str == "slowdown") {
+      action_type_ = SLOWDOWN;
+    } else {  // at_str == "approach"
+      action_type_ = APPROACH;
+    }
+
+    if (action_type_ == STOP) {
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + ".stop_points", rclcpp::ParameterValue(3));
+      stop_points_ = node->get_parameter(polygon_name_ + ".stop_points").as_int();
+    } else if (action_type_ == SLOWDOWN) {
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + ".stop_points", rclcpp::ParameterValue(3));
+      stop_points_ = node->get_parameter(polygon_name_ + ".stop_points").as_int();
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + ".slowdown", rclcpp::ParameterValue(0.5));
+      slowdown_ = node->get_parameter(polygon_name_ + ".slowdown").as_double();
+    } else {  // action_type_ == APPROACH
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + ".time_before_collision",
+        rclcpp::ParameterValue(5.0));
+      time_before_collision_ =
+        node->get_parameter(polygon_name_ + ".time_before_collision").as_double();
+    }
+  } catch (const std::exception & ex) {
+    RCLCPP_ERROR(
+      node->get_logger(), "Error while getting basic polygon parameters: %s", ex.what());
+    return false;
+  }
+
+  return true;
+}
+
+PolygonType PolygonBase::getPolygonType()
+{
+  return polygon_type_;
+}
+
+void PolygonBase::setPolygonType(const PolygonType pt)
+{
+  polygon_type_ = pt;
+}
+
+ActionType PolygonBase::getActionType()
+{
+  return action_type_;
+}
+
+int PolygonBase::getStopPoints()
+{
+  return stop_points_;
+}
+
+void PolygonBase::setStopPoints(const int sp)
+{
+  stop_points_ = sp;
+}
+
+double PolygonBase::getSlowdown()
+{
+  return slowdown_;
+}
+
+void PolygonBase::setSlowdown(const double slowdown)
+{
+  slowdown_ = slowdown;
+}
+
+double PolygonBase::getTimeBeforeCollision()
+{
+  return time_before_collision_;
+}
+
+void PolygonBase::setTimeBeforeCollision(const double tbc)
+{
+  time_before_collision_ = tbc;
+}
+
+void PolygonBase::getCollision(
+  const Point & point, const Velocity & velocity, double & coll_time)
+{
+  // Initial robot pose is {0,0} in base_footprint coordinates
+  Pose pose = {0, 0, 0};
+  Velocity vel = velocity;
+
+  // Robot movement simulation
+  for (double time = 0.0; time <= time_before_collision_; time += simulation_time_step_) {
+    // NOTE: vel is changing during the simulation
+    stepRobot(vel, simulation_time_step_, pose);
+    Point point_fixed = point;
+    transformPoint(pose, point_fixed);
+    if (isPointInside(point_fixed)) {
+      coll_time = time;
+      return;
+    }
+  }
+
+  // There is no collision
+  coll_time = -1.0;
+}
+
+Velocity PolygonBase::getSafeVelocity(
+  const Velocity & velocity, const double coll_time)
```

Comment:
collision_time, don't shorten it, `coll` took me a minute to figure out what that meant

Reply:
[Done](https://github.com/ros-planning/navigation2/pull/2982/files#diff-6ff706b80d78c34ef10bd9efd8109834fe66be34241d395835af7c3857df760eR130-R131)